### PR TITLE
mesa: restore OpenCL support.

### PR DESCRIPTION
**Pull request by athas**
###### Motivation for this change

This is a first shot at addressing #62933.  In practice, OpenCL support should probably be hidden behind an option, because it takes up a significant chunk of space (requires a full LLVM at run-time).

Currently I put all the OpenCL stuff in the `drivers` output, because at least `mesa.icd` needs to be available via `/run/opengl-driver`.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/mesa/default.nix [hunk 1]
```diff
@@ -52,7 +53,7 @@ stdenv.mkDerivation {
   #  ~35 MB in $drivers; watch https://launchpad.net/ubuntu/+source/mesa/+changelog
   patches = [
     ./missing-includes.patch # dev_t needs sys/stat.h, time_t needs time.h, etc.-- fixes build w/musl
-    ./opencl-install-dir.patch
+    ./opencl.patch
```
Comment (killercup) [inline pkgs/development/libraries/mesa/default.nix]: this is missing :)

**Review by killercup — APPROVED** [review]
If someone can tell me how to best test this, I'll give it a go!

Edit: Also cmd+enter triggering the "approve" button is not the wisest choice, github

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (athas) [reply pkgs/development/libraries/mesa/default.nix]: Oops.

**Comment by rnhmjoj** [comment]
@athas As you said, this should be behind a guard like `enableOpenCL`.

I would suggest to create a package `mesa-opencl = mesa.override { enableOpenCL = true; }` or maybe an output like `mesa.opencl`, if possible.
Then just adding it to `opengl.extraPackages` should install the ICD and the driver.

> If someone can tell me how to best test this, I'll give it a go!

If you build mesa from you nixpkgs checkout, you can install it in you system by using the (internal) `opengl.package` option. Alternatively, add it to `opengl.extraPackages` if you only want to test openCL.

To test wheter the openCL driver can be loaded you can use `opengl-info`.
I have a small program that uses an OpenCL kernel to compute the Hurst exponent laying around, I could give you that but I'm sure there are better tests.

**Comment by athas** [comment]
I had trouble using a `mesa.opencl` output, because I got circular dependencies between the `drivers` and `opencl` output.  I think `libOpenCL.so` (which is big) *has* to live in `drivers`, because that is where `mesa.icd` must be in order to be accessible via `/run/opengl-driver/etc/OpenCL/mesa.icd`.  At least, addressing this is beyond my current skill set.

For the same reason, it's also a little tricky to debug, since the ICD loader only seems to pick up stuff in `/run/opengl-driver/etc/OpenCL`.

**Comment by killercup** [comment]
> you can install it in you system by using the (internal) opengl.package option. Alternatively, add it to opengl.extraPackages if you only want to test openCL.

Thanks for the tips, @rnhmjoj! This works for me with a 5700XT:

```nix
{
hardware.opengl = let
    opencl_pr = import (builtins.fetchTarball {
        name = "opencl_pr";
        url = "https://github.com/athas/nixpkgs/archive/f92a2a9b69eba9909d25ffaab6ded4d6f0f4efad.tar.gz";
    }) { };
in {
    enable = true;
    driSupport32Bit = true;
    package = opencl_pr.mesa.drivers;
    extraPackages = [ opencl_pr.mesa ];
};
}
```

`clinfo` output looks good, but it also contains

```console
=== CL_PROGRAM_BUILD_LOG ===
fatal error: cannot open file '/nix/store/l81x3yfwahahhxmfk4wbg2ki025v451p-libclc-2019-06-09/lib/clc/gfx1010-amdgcn-mesa-mesa3d.bc': No such file or directory
  Preferred work group size multiple              <getWGsizes:1200: create kernel : error -46>
```

Not sure if I should worry about that.

**Comment by athas** [comment]
I think the `libclc` is just too old for that GPU.  I'll make a PR about bumping `libclc`.

**Comment by athas** [comment]
Oh, apparently Mesa OpenCL doesn't support Navi cards at all...

**Comment by killercup** [comment]
Oh, that's too bad! This is what clinfo tell me now, but I don't really have a clue what everything means.

```console
$ clinfo                                                                                                                                                                                                                                                   
Number of platforms                               1
  Platform Name                                   Clover
  Platform Vendor                                 Mesa
  Platform Version                                OpenCL 1.1 Mesa 19.3.3
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_icd
  Platform Extensions function suffix             MESA

  Platform Name                                   Clover
Number of devices                                 1
  Device Name                                     AMD NAVI10 (DRM 3.36.0, 5.5.9, LLVM 9.0.1)
  Device Vendor                                   AMD
  Device Vendor ID                                0x1002
  Device Version                                  OpenCL 1.1 Mesa 19.3.3
  Driver Version                                  19.3.3
  Device OpenCL C Version                         OpenCL C 1.1 
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Max compute units                               40
  Max clock frequency                             2100MHz
  Max work item dimensions                        3
  Max work item sizes                             256x256x256
  Max work group size                             256
=== CL_PROGRAM_BUILD_LOG ===
fatal error: cannot open file '/nix/store/l81x3yfwahahhxmfk4wbg2ki025v451p-libclc-2019-06-09/lib/clc/gfx1010-amdgcn-mesa-mesa3d.bc': No such file or directory
  Preferred work group size multiple              <getWGsizes:1200: create kernel : error -46>
  Preferred / native vector sizes                 
    char                                                16 / 16      
    short                                                8 / 8       
    int                                                  4 / 4       
    long                                                 2 / 2       
    half                                                 8 / 8        (cl_khr_fp16)
    float                                                4 / 4       
    double                                               2 / 2        (cl_khr_fp64)
  Half-precision Floating-point support           (cl_khr_fp16)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
  Single-precision Floating-point support         (core)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    64, Little-Endian
  Global memory size                              8589934592 (8GiB)
  Error Correction support                        No
  Max memory allocation                           6871947673 (6.4GiB)
  Unified memory for Host and Device              No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       32768 bits (4096 bytes)
  Global Memory cache type                        None
  Image support                                   No
  Local memory type                               Local
  Local memory size                               32768 (32KiB)
  Max number of constant args                     16
  Max constant buffer size                        2147483647 (2GiB)
  Max size of kernel argument                     1024
  Queue properties                                
    Out-of-order execution                        No
    Profiling                                     Yes
  Profiling timer resolution                      0ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
  Device Extensions                               cl_khr_byte_addressable_store cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_fp64 cl_khr_fp16

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  Clover
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [MESA]
  clCreateContext(NULL, ...) [default]            Success [MESA]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (1)
    Platform Name                                 Clover
    Device Name                                   AMD NAVI10 (DRM 3.36.0, 5.5.9, LLVM 9.0.1)
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (1)
    Platform Name                                 Clover
    Device Name                                   AMD NAVI10 (DRM 3.36.0, 5.5.9, LLVM 9.0.1)
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (1)
    Platform Name                                 Clover
    Device Name                                   AMD NAVI10 (DRM 3.36.0, 5.5.9, LLVM 9.0.1)

ICD loader properties
  ICD loader Name                                 OpenCL ICD Loader
  ICD loader Vendor                               OCL Icd free software
  ICD loader Version                              2.2.10
  ICD loader Profile                              OpenCL 2.1
```

**Comment by athas** [comment]
I think that means that all the infrastructure works.  The OpenCL ICD loader can find the Mesa ICD and talk to the hardware.  It just can't compile any kernels, because some GPU-specific library files are missing.

[Here](https://gist.github.com/athas/8a169f818eb768d8d5a588d4a4f1eaf4) is a small OpenCL test program that compiles and launches a kernel.  Compile with

```
$ nix-shell -p ocl-icd -p opencl-headers --run 'gcc rot13.c -o rot13 -lOpenCL'
```

and then run with

```
$ ./rot13
```

I don't expect it'll work on your machine.  It also doesn't work on my Vega 64, as the kernel launch hangs forever.  That's an improvement, though: Mesa OpenCL used to reboot my computer, even on non-NixOS distributions.

The general buggyness of Mesa OpenCL is going to be an obstacle to testing this...

**Comment by rnhmjoj** [comment]
> Oh, that's too bad! This is what clinfo tell me now, but I don't really have a clue what everything means.

That looks good. In case the driver is not found you would simply get:

    ERROR: clGetPlatformIDs(-1001)

**Comment by killercup** [comment]
@athas yep, doesn't start, "fatal error: cannot open file '/nix/store/l81x3yfwahahhxmfk4wbg2ki025v451p-libclc-2019-06-09/lib/clc/gfx1010-amdgcn-mesa-mesa3d.bc': No such file or directory".

**Comment by volth** [comment]
OpenCL support as in was around 19.03 allowed to use both Intel GPU and Nvidia GPU on laptops with hybrid graphics.

Not everything is broken, but this patch does not restore old style support.

With this patch, either Intel GPU or Nvidia GPU is available depending on the BIOS setting (with "hybrid graphics" only Intel GPU is available to OpenCL, with "discrete graphics" - only Nvidia GPU)

Of course it is a big step forward comparing to the current broken state, but it could be improved

**Comment by davidak** [comment]
I added the code from https://github.com/NixOS/nixpkgs/pull/82729#issuecomment-600101086 to my `configuration.nix` and now OpenCL works!

```
[davidak@gaming:~]$ nix run nixpkgs.clinfo -c clinfo
Number of platforms                               1
  Platform Name                                   Clover
  Platform Vendor                                 Mesa
  Platform Version                                OpenCL 1.1 Mesa 19.3.3
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_icd
  Platform Extensions function suffix             MESA

  Platform Name                                   Clover
Number of devices                                 1
  Device Name                                     AMD Radeon (TM) RX 480 Graphics (POLARIS10, DRM 3.35.0, 5.4.41, LLVM 9.0.1)
```

Even BOINC detects the GPU now:

Screenshot from 2020-05-17 00-05-40

**Review by davidak — APPROVED** [review]
works, changes looks OK as far as i can judge

**Comment by rnhmjoj** [comment]
@vcunat can you take a look at this PR?

**Review by rnhmjoj — CHANGES_REQUESTED** [review]
The change looks good but openCL support should be optional to avoid a large closure size increase.
For convenience, a `mesa-opencl` package could be added so that users don't have to rebuild mesa.

**Comment by cab404** [comment]
I tried it, but blender doesn't report my videocard (540X) as a usable device. Is that normal?

**Comment by davidak** [comment]
@cab404 i also don't see my AMD Radeon RX 480 as supported. so might be blender specific

**Comment by mohe2015** [comment]
@athas May you rebase this please?

**Comment by athas** [comment]
I'll try over the weekend, but I won't have the spare life cycles to actually test it.

**Comment by mohe2015** [comment]
Cool @athas. https://github.com/mohe2015/nixpkgs/tree/mesa-opencl may or may not be a correct rebase (nonsquashed)

**Comment by SuperSandro2000** [comment]
Can you please fix the merge conflict and target staging with the mass rebuild?

**Comment by timhae** [comment]
I also added the code from [the comment](https://github.com/NixOS/nixpkgs/pull/82729#issuecomment-600101086) to my `configuration.nix` and was able to compile the program mentioned in the [other comment](https://github.com/NixOS/nixpkgs/pull/82729#issuecomment-600234320), the output is (as expected) `Hello, World!`. Thanks a lot for your effort and please let me know if I can do any additional testing to help you.

**Comment by athas** [comment]
I have rebased on top of master.  Haven't tried running it, but by inspection of the files, it looks OK.

File: pkgs/development/libraries/mesa/default.nix [hunk 2]
```diff
@@ -176,9 +182,22 @@ stdenv.mkDerivation {
 
     if [ -n "$(shopt -s nullglob; echo "$out"/lib/lib*_mesa*)" ]; then
       # Move other drivers to a separate output
-      mv $out/lib/lib*_mesa* $drivers/lib
+      mv -t $drivers/lib \
+        $out/lib/lib*_mesa*
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/mesa/default.nix:186]:
```suggestion
      mv -t $drivers/lib $out/lib/lib*_mesa*
```

File: pkgs/development/libraries/mesa/default.nix [hunk 3]
```diff
@@ -221,7 +240,10 @@ stdenv.mkDerivation {
     done
   '';
 
-  NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.isDarwin "-fno-common";
+  NIX_CFLAGS_COMPILE =
+    optionals stdenv.isDarwin ["-fno-common"]
+    ++ [ "-UPIPE_SEARCH_DIR"
+         "-DPIPE_SEARCH_DIR=\"${placeholder "drivers"}/lib/gallium-pipe\"" ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/mesa/default.nix:246]:
```suggestion
  NIX_CFLAGS_COMPILE = optionals stdenv.isDarwin ["-fno-common"] ++ [ 
    "-UPIPE_SEARCH_DIR"
    "-DPIPE_SEARCH_DIR=\"${placeholder "drivers"}/lib/gallium-pipe\"" 
  ];
```

**Comment by SuperSandro2000** [comment]
Please target staging and squash the commits.

**Comment by shpinog** [comment]
I also use the workaround above. Unfortunately, rocm does not work on many platforms, as it requires hardware support for the processor.

File: pkgs/development/libraries/mesa/default.nix [hunk 4]
```diff
@@ -176,9 +182,21 @@ stdenv.mkDerivation {
...
 
+    if [ -n "$(shopt -s nullglob; echo "$out"/lib/libMesaOpenCL*)" ]; then
+      # Move OpenCL stuff
+      mv -t "$drivers/lib/"     \
+        $out/lib/gallium-pipe   \
+        $out/lib/libMesaOpenCL*
+    fi
+
+    # We construct our own .icd file that contains an absolute path.
+    rm -rf $out/etc/OpenCL
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/mesa/default.nix:196]:
```suggestion
    rm -r $out/etc/OpenCL
```

**Comment by davidak** [comment]
I also used this workaround, since may last year and noticed now that it does not work anymore.

My config.: https://codeberg.org/davidak/nixos-config/src/commit/3d15f77fbe04177c546c0a15026ec1f3240d50ac/modules/amd/default.nix

I don't see any OpenCL device:

```
[davidak@gaming:~/code/nixpkgs]$ nix run nixpkgs.clinfo -c clinfo
Number of platforms                               0
```

- system: `"x86_64-linux"`
- host os: `Linux 5.11.2, NixOS, 21.05pre274338.102eb68ceec (Okapi)`
- multi-user?: `yes`
- sandbox: `yes`
- version: `nix-env (Nix) 2.3.10`
- channels(root): `"home-manager-20.09, nixos-21.05pre279456.04a2b269d89, nixos-hardware, nixos-unstable-21.05pre279456.04a2b269d89"`
- nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maybe this needs an update?

**Comment by kvtb** [comment]
Could we merge it under `enableOpelCL` conditional, so those who concern the closure size could opt-out?
It is the only showstopper, isn't it?

**Comment by sbruder** [comment]
I rebased the PR (and the included patch against mesa) onto the latest master and added a switch for making OpenCL support optional (as @rnhmjoj suggested): https://github.com/sbruder/nixpkgs/commits/mesa-opencl and tested the PR [cherry-picked onto 21.05](https://github.com/sbruder/nixpkgs/commits/mesa-opencl-21.05) with the following config:
```nix
{ pkgs, ... }:
{
  hardware.opengl =
    let
      pkgsMesaClover = import
        (pkgs.fetchFromGitHub {
          owner = "sbruder";
          repo = "nixpkgs";
          rev = "41d33334ec2d07da8f5bd0f3749949241c42266b";
          sha256 = "sha256-sA/kwZ8kXuRC/sLjja2ISpyTnlEwZjifyYeiz9hFWRE=";
        })
        { inherit (pkgs) system; };
    in
    {
      package = pkgsMesaClover.mesa-opencl.drivers;
      extraPackages = [
        pkgsMesaClover.mesa-opencl
      ];
    };
}
```

After a rebuild `clinfo` detects my RX 480 ([gist with the output of clinfo and comparison to its output with ROCm](https://gist.github.com/sbruder/8df49af6362b6979205a0f18dac923c7)), though it only shows support for OpenCL 1.1, not 1.2 [as it should support](https://www.phoronix.com/scan.php?page=news_item&px=Mesa-20.3-OpenCL-1.2-Clover) (though I don’t know if upstream does not support it on my hardware or if this is due to how nixpkgs packages it).

Some programs that work include @athas’ rot13 example, hashcat (with `--force`) and waifu2x-converter-cpp. Others that don’t work include Blender (which is a [known problem](https://blender.stackexchange.com/q/221683) and [Blender will drop OpenCL support in the near future](https://blender.stackexchange.com/a/221710)) and the NNEDI3CL VapourSynth plugin (fails with `Boost.Compute: kernel compilation failed (-43)`).

Besides that, some programs that do work show warnings like this, which probably are not problematic, but seems like it’s not fully working yet:
```
'+fp32-denormals' is not a recognized feature for this target (ignoring feature)
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'+fp32-denormals' is not a recognized feature for this target (ignoring feature)
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'+fp32-denormals' is not a recognized feature for this target (ignoring feature)
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'+fp32-denormals' is not a recognized feature for this target (ignoring feature)
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
```

**Comment by rnhmjoj** [comment]
@sbruder can you make a new pull request (against staging) with your patch?

**Comment by primeos** [comment]
I'll go ahead and close this in favor of #136402 (will be merged soon).

Thanks @athas et al.

**Pull request closed without merging**